x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,488 @@
+
+            self.do_io(PollMode::ReadAndWritable, |conn| {
+                conn.inner.extension_information(name)
+            })
+            .await
+        })
+    }
+
+    fn maximum_request_bytes(&self) -> Pin<Box<dyn Future<Output = usize> + Send + '_>> {
+        Box::pin(async move {
+            // Gain exclusive access to both ends.
+            let _guard1 = self.write_lock.lock().await;
+            let _guard2 = self.read_lock.lock().await;
+
+            // No real way to do this asynchronously.
+            let inner = self.inner.clone();
+            blocking::unblock(move || inner.maximum_request_bytes()).await
+        })
+    }
+
+    fn prefetch_maximum_request_bytes(&self) -> Pin<Box<dyn Future<Output = ()> + Send + '_>> {
+        Box::pin(async move {
+            // Gain exclusive access to the write end.
+            let _guard1 = self.write_lock.lock().await;
+
+            // No real way to do this asynchronously.
+            let inner = self.inner.clone();
+            blocking::unblock(move || inner.prefetch_maximum_request_bytes()).await
+        })
+    }
+
+    fn wait_for_reply(
+        &self,
+        sequence: SequenceNumber,
+    ) -> Fut<'_, Option<Self::Buf>, ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the read end of the stream.
+            let _guard = self.read_lock.lock().await;
+
+            self.do_io(PollMode::Readable, |conn| {
+                conn.inner.wait_for_reply(sequence)
+            })
+            .await
+        })
+    }
+
+    fn wait_for_reply_or_raw_error(
+        &self,
+        sequence: SequenceNumber,
+    ) -> Fut<'_, ReplyOrError<Self::Buf>, ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the read end of the stream.
+            let _guard = self.read_lock.lock().await;
+
+            self.do_io(PollMode::Readable, |conn| {
+                conn.inner.wait_for_reply_or_raw_error(sequence)
+            })
+            .await
+        })
+    }
+
+    fn wait_for_reply_with_fds_raw(
+        &self,
+        sequence: SequenceNumber,
+    ) -> Fut<'_, ReplyOrError<x11rb::connection::BufWithFds<Self::Buf>, Self::Buf>, ConnectionError>
+    {
+        Box::pin(async move {
+            // Gain exclusive access to the read end of the stream.
+            let _guard = self.read_lock.lock().await;
+
+            self.do_io(PollMode::Readable, |conn| {
+                conn.inner.wait_for_reply_with_fds_raw(sequence)
+            })
+            .await
+        })
+    }
+}
+
+impl<S: Stream + Send + Sync + 'static> Connection for RustConnection<S> {
+    fn poll_for_raw_event(
+        &self,
+    ) -> Result<Option<RawEventAndSeqNumber<Self::Buf>>, ConnectionError> {
+        // Try to gain access to the read end of the stream.
+        let _guard = match self.read_lock.try_lock() {
+            Some(guard) => guard,
+            None => return Ok(None),
```

Comment:
Couldn't this incorrectly return `None` when an event sits in a queue / was already received, but the mutex is locked because something waits for the reply to a request?